// @nhost/hasura-auth-js@2.1.8/dist/index.esm.js
import ce from "jwt-decode";
import { createMachine as P, send as p, assign as E, interpret as b } from "xstate";
import L from "js-cookie";
import ue from "fetch-ponyfill";
const O = "nhostRefreshToken", N = "nhostRefreshTokenId", R = "nhostRefreshTokenExpiresAt", le = 3, de = 60, H = 5, Z = 0, J = 1, y = 10, C = 20;
class U extends Error {
  constructor(e) {
    super(e.message), Error.captureStackTrace && Error.captureStackTrace(this, this.constructor), e instanceof Error ? (this.name = e.name, this.error = {
      error: e.name,
      status: J,
      message: e.message
    }) : (this.name = e.error, this.error = e);
  }
}
const I = {
  status: y,
  error: "invalid-email",
  message: "Email is incorrectly formatted"
}, fe = {
  status: y,
  error: "invalid-mfa-type",
  message: "MFA type is invalid"
}, he = {
  status: y,
  error: "invalid-mfa-code",
  message: "MFA code is invalid"
}, $ = {
  status: y,
  error: "invalid-password",
  message: "Password is incorrectly formatted"
}, Y = {
  status: y,
  error: "invalid-phone-number",
  message: "Phone number is incorrectly formatted"
}, me = {
  status: y,
  error: "invalid-mfa-ticket",
  message: "MFA ticket is invalid"
}, Ee = {
  status: y,
  error: "no-mfa-ticket",
  message: "No MFA ticket has been provided"
}, ge = {
  status: y,
  error: "no-refresh-token",
  message: "No refresh token has been provided"
}, pe = {
  status: C,
  error: "refresher-already-running",
  message: "The token refresher is already running. You must wait until is has finished before submitting a new token."
}, S = {
  status: C,
  error: "already-signed-in",
  message: "User is already signed in"
}, Te = {
  status: C,
  error: "unauthenticated-user",
  message: "User is not authenticated"
}, or = {
  status: C,
  error: "user-not-anonymous",
  message: "User is not anonymous"
}, we = {
  status: C,
  error: "unverified-user",
  message: "Email needs verification"
}, ye = {
  status: y,
  error: "invalid-refresh-token",
  message: "Invalid or expired refresh token"
}, Se = {
  status: J,
  error: "invalid-sign-in-method",
  message: "Invalid sign-in method"
}, G = {
  user: null,
  mfa: null,
  accessToken: {
    value: null,
    expiresAt: null,
    expiresInSeconds: 15
  },
  refreshTimer: {
    startedAt: null,
    attempts: 0,
    lastAttempt: null
  },
  refreshToken: {
    value: null
  },
  importTokenAttempts: 0,
  errors: {}
};
function ke(n) {
  return new TextEncoder().encode(n);
}
function _(n) {
  const e = new Uint8Array(n);
  let r = "";
  for (const s of e)
    r += String.fromCharCode(s);
  return btoa(r).replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}
function j(n) {
  const e = n.replace(/-/g, "+").replace(/_/g, "/"), r = (4 - e.length % 4) % 4, t = e.padEnd(e.length + r, "="), s = atob(t), i = new ArrayBuffer(s.length), u = new Uint8Array(i);
  for (let d = 0; d < s.length; d++)
    u[d] = s.charCodeAt(d);
  return i;
}
function ee() {
  return (window == null ? void 0 : window.PublicKeyCredential) !== void 0 && typeof window.PublicKeyCredential == "function";
}
function re(n) {
  const { id: e } = n;
  return {
    ...n,
    id: j(e),
    transports: n.transports
  };
}
function te(n) {
  return n === "localhost" || /^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$/i.test(n);
}
class g extends Error {
  constructor(e, r = "WebAuthnError") {
    super(e), this.name = r;
  }
}
function Ae({ error: n, options: e }) {
  var r, t;
  const { publicKey: s } = e;
  if (!s)
    throw Error("options was missing required publicKey property");
  if (n.name === "AbortError") {
    if (e.signal === new AbortController().signal)
      return new g("Registration ceremony was sent an abort signal", "AbortError");
  } else if (n.name === "ConstraintError") {
    if (((r = s.authenticatorSelection) === null || r === void 0 ? void 0 : r.requireResidentKey) === !0)
      return new g("Discoverable credentials were required but no available authenticator supported it", "ConstraintError");
    if (((t = s.authenticatorSelection) === null || t === void 0 ? void 0 : t.userVerification) === "required")
      return new g("User verification was required but no available authenticator supported it", "ConstraintError");
  } else {
    if (n.name === "InvalidStateError")
      return new g("The authenticator was previously registered", "InvalidStateError");
    if (n.name === "NotAllowedError")
      return new g("User clicked cancel, or the registration ceremony timed out", "NotAllowedError");
    if (n.name === "NotSupportedError")
      return s.pubKeyCredParams.filter((u) => u.type === "public-key").length === 0 ? new g('No entry in pubKeyCredParams was of type "public-key"', "NotSupportedError") : new g("No available authenticator supported any of the specified pubKeyCredParams algorithms", "NotSupportedError");
    if (n.name === "SecurityError") {
      const i = window.location.hostname;
      if (te(i)) {
        if (s.rp.id !== i)
          return new g(`The RP ID "${s.rp.id}" is invalid for this domain`, "SecurityError");
      } else
        return new g(`${window.location.hostname} is an invalid domain`, "SecurityError");
    } else if (n.name === "TypeError") {
      if (s.user.id.byteLength < 1 || s.user.id.byteLength > 64)
        return new g("User ID was not between 1 and 64 characters", "TypeError");
    } else if (n.name === "UnknownError")
      return new g("The authenticator was unable to process the specified options, or could not create a new credential", "UnknownError");
  }
  return n;
}
class Ie {
  createNewAbortSignal() {
    return this.controller && this.controller.abort(), this.controller = new AbortController(), this.controller.signal;
  }
  reset() {
    this.controller = void 0;
  }
}
const K = new Ie();
async function ne(n) {
  if (!ee())
    throw new Error("WebAuthn is not supported in this browser");
  const r = { publicKey: {
    ...n,
    challenge: j(n.challenge),
    user: {
      ...n.user,
      id: ke(n.user.id)
    },
    excludeCredentials: n.excludeCredentials.map(re)
  } };
  r.signal = K.createNewAbortSignal();
  let t;
  try {
    t = await navigator.credentials.create(r);
  } catch (h) {
    throw Ae({ error: h, options: r });
  } finally {
    K.reset();
  }
  if (!t)
    throw new Error("Registration was not completed");
  const { id: s, rawId: i, response: u, type: d } = t, f = {
    id: s,
    rawId: _(i),
    response: {
      attestationObject: _(u.attestationObject),
      clientDataJSON: _(u.clientDataJSON)
    },
    type: d,
    clientExtensionResults: t.getClientExtensionResults(),
    authenticatorAttachment: t.authenticatorAttachment
  };
  return typeof u.getTransports == "function" && (f.transports = u.getTransports()), f;
}
function ve(n) {
  return new TextDecoder("utf-8").decode(n);
}
async function _e() {
  if (navigator.credentials.conditionalMediationSupported)
    return !0;
  const n = window.PublicKeyCredential;
  return n.isConditionalMediationAvailable !== void 0 && n.isConditionalMediationAvailable();
}
function Re({ error: n, options: e }) {
  var r;
  const { publicKey: t } = e;
  if (!t)
    throw Error("options was missing required publicKey property");
  if (n.name === "AbortError") {
    if (e.signal === new AbortController().signal)
      return new g("Authentication ceremony was sent an abort signal", "AbortError");
  } else {
    if (n.name === "NotAllowedError")
      return !((r = t.allowCredentials) === null || r === void 0) && r.length ? new g("No available authenticator recognized any of the allowed credentials", "NotAllowedError") : new g("User clicked cancel, or the authentication ceremony timed out", "NotAllowedError");
    if (n.name === "SecurityError") {
      const s = window.location.hostname;
      if (te(s)) {
        if (t.rpId !== s)
          return new g(`The RP ID "${t.rpId}" is invalid for this domain`, "SecurityError");
      } else
        return new g(`${window.location.hostname} is an invalid domain`, "SecurityError");
    } else if (n.name === "UnknownError")
      return new g("The authenticator was unable to process the specified options, or could not create a new assertion signature", "UnknownError");
  }
  return n;
}
async function Pe(n, e = !1) {
  var r, t;
  if (!ee())
    throw new Error("WebAuthn is not supported in this browser");
  let s;
  ((r = n.allowCredentials) === null || r === void 0 ? void 0 : r.length) !== 0 && (s = (t = n.allowCredentials) === null || t === void 0 ? void 0 : t.map(re));
  const i = {
    ...n,
    challenge: j(n.challenge),
    allowCredentials: s
  }, u = {};
  if (e) {
    if (!await _e())
      throw Error("Browser does not support WebAuthn autofill");
    if (document.querySelectorAll("input[autocomplete*='webauthn']").length < 1)
      throw Error('No <input> with `"webauthn"` in its `autocomplete` attribute was detected');
    u.mediation = "conditional", i.allowCredentials = [];
  }
  u.publicKey = i, u.signal = K.createNewAbortSignal();
  let d;
  try {
    d = await navigator.credentials.get(u);
  } catch (l) {
    throw Re({ error: l, options: u });
  } finally {
    K.reset();
  }
  if (!d)
    throw new Error("Authentication was not completed");
  const { id: f, rawId: h, response: a, type: o } = d;
  let c;
  return a.userHandle && (c = ve(a.userHandle)), {
    id: f,
    rawId: _(h),
    response: {
      authenticatorData: _(a.authenticatorData),
      clientDataJSON: _(a.clientDataJSON),
      signature: _(a.signature),
      userHandle: c
    },
    type: o,
    clientExtensionResults: d.getClientExtensionResults(),
    authenticatorAttachment: d.authenticatorAttachment
  };
}
const V = typeof window != "undefined", M = /* @__PURE__ */ new Map(), Oe = (n) => {
  var e;
  return V && typeof localStorage != "undefined" ? localStorage.getItem(n) : (e = M.get(n)) != null ? e : null;
}, be = (n, e) => {
  V && typeof localStorage != "undefined" ? e ? localStorage.setItem(n, e) : localStorage.removeItem(n) : e ? M.set(n, e) : M.has(n) && M.delete(n);
}, Ce = (n, e) => {
  if (n === "localStorage" || n === "web")
    return Oe;
  if (n === "cookie")
    return (r) => {
      var t;
      return V && (t = L.get(r)) != null ? t : null;
    };
  if (!e)
    throw Error(
      `clientStorageType is set to '${n}' but no clientStorage has been given`
    );
  if (n === "react-native")
    return (r) => {
      var t;
      return (t = e.getItem) == null ? void 0 : t.call(e, r);
    };
  if (n === "capacitor")
    return (r) => {
      var t;
      return (t = e.get) == null ? void 0 : t.call(e, { key: r });
    };
  if (n === "expo-secure-storage")
    return (r) => {
      var t;
      return (t = e.getItemAsync) == null ? void 0 : t.call(e, r);
    };
  if (n === "custom") {
    if (e.getItem && e.removeItem)
      return e.getItem;
    if (e.getItemAsync)
      return e.getItemAsync;
    throw Error(
      `clientStorageType is set to 'custom' but clientStorage is missing either "getItem" and "removeItem" properties or "getItemAsync" property`
    );
  }
  throw Error(`Unknown storage type: ${n}`);
}, Ne = (n, e) => {
  if (n === "localStorage" || n === "web")
    return be;
  if (n === "cookie")
    return (r, t) => {
      V && (t ? L.set(r, t, { expires: 30, sameSite: "lax", httpOnly: !1 }) : L.remove(r));
    };
  if (!e)
    throw Error(
      `clientStorageType is set to '${n}' but no clienStorage has been given`
    );
  if (n === "react-native")
    return (r, t) => {
      var s, i;
      return t ? (s = e.setItem) == null ? void 0 : s.call(e, r, t) : (i = e.removeItem) == null ? void 0 : i.call(e, r);
    };
  if (n === "capacitor")
    return (r, t) => {
      var s, i;
      return t ? (s = e.set) == null ? void 0 : s.call(e, { key: r, value: t }) : (i = e.remove) == null ? void 0 : i.call(e, { key: r });
    };
  if (n === "expo-secure-storage")
    return async (r, t) => {
      var s, i;
      return t ? (s = e.setItemAsync) == null ? void 0 : s.call(e, r, t) : (i = e.deleteItemAsync) == null ? void 0 : i.call(e, r);
    };
  if (n === "custom") {
    if (!e.removeItem)
      throw Error(
        "clientStorageType is set to 'custom' but clientStorage is missing a removeItem property"
      );
    if (e.setItem)
      return (r, t) => {
        var s, i;
        return t ? (s = e.setItem) == null ? void 0 : s.call(e, r, t) : (i = e.removeItem) == null ? void 0 : i.call(e, r);
      };
    if (e.setItemAsync)
      return async (r, t) => {
        var s, i;
        return t ? (s = e.setItemAsync) == null ? void 0 : s.call(e, r, t) : (i = e.removeItem) == null ? void 0 : i.call(e, r);
      };
    throw Error(
      "clientStorageType is set to 'custom' but clientStorage is missing setItem or setItemAsync property"
    );
  }
  throw Error(`Unknown storage type: ${n}`);
}, x = (n) => !n || !n.accessToken.value || !n.refreshToken.value || !n.accessToken.expiresAt || !n.user ? null : {
  accessToken: n.accessToken.value,
  accessTokenExpiresIn: (n.accessToken.expiresAt.getTime() - Date.now()) / 1e3,
  refreshToken: n.refreshToken.value,
  user: n.user
}, A = ({
  accessToken: n,
  isError: e,
  user: r,
  error: t
}) => e ? {
  session: null,
  error: t
} : r && n ? {
  // TODO either return the refresh token or remove it from the session type
  session: { accessToken: n, accessTokenExpiresIn: 0, refreshToken: "", user: r },
  error: null
} : { session: null, error: null }, q = () => typeof window != "undefined", { fetch: xe } = ue(), se = async (n, e, { token: r, body: t } = {}) => {
  const s = {
    "Content-Type": "application/json",
    Accept: "*/*"
  };
  r && (s.Authorization = `Bearer ${r}`);
  const i = {
    method: e,
    headers: s
  };
  t && (i.body = JSON.stringify(t));
  try {
    const u = await xe(n, i);
    if (!u.ok) {
      const d = await u.json();
      return Promise.reject({ error: d });
    }
    try {
      return { data: await u.json(), error: null };
    } catch {
      return console.warn(`Unexpected response: can't parse the response of the server at ${n}`), { data: "OK", error: null };
    }
  } catch {
    const d = {
      message: "Network Error",
      status: Z,
      error: "network"
    };
    return Promise.reject({ error: d });
  }
}, k = async (n, e, r) => se(n, "POST", { token: r, body: e }), De = (n, e) => se(n, "GET", { token: e }), ie = (n, e) => {
  const r = e && Object.entries(e).map(([t, s]) => {
    const i = Array.isArray(s) ? s.join(",") : typeof s == "object" ? JSON.stringify(s) : s;
    return `${t}=${encodeURIComponent(i)}`;
  }).join("&");
  return r ? `${n}?${r}` : n;
}, w = (n, e) => {
  if (!(e != null && e.redirectTo))
    return e;
  const { redirectTo: r, ...t } = e;
  if (!n)
    return r.startsWith("/") ? t : e;
  const s = new URL(n), i = Object.fromEntries(new URLSearchParams(s.search)), u = new URL(r.startsWith("/") ? s.origin + r : r), d = new URLSearchParams(u.search);
  let f = Object.fromEntries(d);
  r.startsWith("/") && (f = { ...i, ...f });
  let h = s.pathname;
  return u.pathname.length > 1 && (h += u.pathname.slice(1)), {
    ...t,
    redirectTo: ie(u.origin + h, f)
  };
};
function D(n, e) {
  var s;
  if (!e) {
    if (typeof window == "undefined")
      return;
    e = ((s = window.location) == null ? void 0 : s.href) || "";
  }
  n = n.replace(/[\[\]]/g, "\\$&");
  const r = new RegExp("[?&#]" + n + "(=([^&#]*)|&|#|$)"), t = r.exec(e);
  return t ? t[2] ? decodeURIComponent(t[2].replace(/\+/g, " ")) : "" : null;
}
function F(n) {
  var r;
  if (typeof window == "undefined")
    return;
  const e = window == null ? void 0 : window.location;
  if (e && e) {
    const t = new URLSearchParams(e.search), s = new URLSearchParams((r = e.hash) == null ? void 0 : r.slice(1));
    t.delete(n), s.delete(n);
    let i = window.location.pathname;
    Array.from(t).length && (i += `?${t.toString()}`), Array.from(s).length && (i += `#${s.toString()}`), window.history.pushState({}, "", i);
  }
}
const v = (n) => !!n && typeof n == "string" && !!String(n).toLowerCase().match(
  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
), W = (n) => !!n && typeof n == "string" && n.length >= le, z = (n) => !!n && typeof n == "string", Ue = (n) => n && typeof n == "string" && n.match(/^mfaTotp:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i), Me = ({
  backendUrl: n,
  clientUrl: e,
  clientStorageType: r = "web",
  clientStorage: t,
  refreshIntervalTime: s,
  autoRefreshToken: i = !0,
  autoSignIn: u = !0
}) => {
  const d = Ce(r, t), f = Ne(r, t), h = async (a, o, c) => (await k(`${n}${a}`, o, c)).data;
  return P(
    {
      schema: {
        context: {},
        events: {},
        services: {}
      },
      tsTypes: {},
      context: G,
      predictableActionArguments: !0,
      id: "nhost",
      type: "parallel",
      states: {
        authentication: {
          initial: "starting",
          on: {
            SESSION_UPDATE: [
              {
                cond: "hasSession",
                actions: ["saveSession", "resetTimer", "reportTokenChanged"],
                target: ".signedIn"
              }
            ]
          },
          states: {
            starting: {
              tags: ["loading"],
              always: { cond: "isSignedIn", target: "signedIn" },
              invoke: {
                id: "importRefreshToken",
                src: "importRefreshToken",
                onDone: [
                  {
                    cond: "hasSession",
                    actions: ["saveSession", "reportTokenChanged"],
                    target: "signedIn"
                  },
                  {
                    target: "signedOut"
                  }
                ],
                onError: [
                  {
                    cond: "shouldRetryImportToken",
                    actions: "incrementTokenImportAttempts",
                    target: "retryTokenImport"
                  },
                  { actions: ["saveAuthenticationError"], target: "signedOut" }
                ]
              }
            },
            retryTokenImport: {
              tags: ["loading"],
              after: {
                RETRY_IMPORT_TOKEN_DELAY: "starting"
              }
            },
            signedOut: {
              initial: "noErrors",
              entry: "reportSignedOut",
              states: {
                noErrors: {},
                success: {},
                needsSmsOtp: {},
                needsMfa: {},
                failed: {},
                signingOut: {
                  entry: ["clearContextExceptRefreshToken"],
                  exit: ["destroyRefreshToken", "reportTokenChanged"],
                  invoke: {
                    src: "signout",
                    id: "signingOut",
                    onDone: {
                      target: "success"
                    },
                    onError: {
                      target: "failed",
                      actions: ["saveAuthenticationError"]
                    }
                  }
                }
              },
              on: {
                SIGNIN_PASSWORD: "authenticating.password",
                SIGNIN_ANONYMOUS: "authenticating.anonymous",
                SIGNIN_SECURITY_KEY_EMAIL: "authenticating.securityKeyEmail",
                SIGNIN_MFA_TOTP: "authenticating.mfa.totp",
                SIGNIN_PAT: "authenticating.pat"
              }
            },
            authenticating: {
              entry: "resetErrors",
              states: {
                password: {
                  invoke: {
                    src: "signInPassword",
                    id: "authenticateUserWithPassword",
                    onDone: [
                      {
                        cond: "hasMfaTicket",
                        actions: ["saveMfaTicket"],
                        target: "#nhost.authentication.signedOut.needsMfa"
                      },
                      {
                        actions: ["saveSession", "reportTokenChanged"],
                        target: "#nhost.authentication.signedIn"
                      }
                    ],
                    onError: [
                      {
                        cond: "unverified",
                        target: [
                          "#nhost.authentication.signedOut",
                          "#nhost.registration.incomplete.needsEmailVerification"
                        ]
                      },
                      {
                        actions: "saveAuthenticationError",
                        target: "#nhost.authentication.signedOut.failed"
                      }
                    ]
                  }
                },
                pat: {
                  invoke: {
                    src: "signInPAT",
                    id: "authenticateWithPAT",
                    onDone: {
                      actions: ["savePATSession", "reportTokenChanged"],
                      target: "#nhost.authentication.signedIn"
                    },
                    onError: {
                      actions: "saveAuthenticationError",
                      target: "#nhost.authentication.signedOut.failed"
                    }
                  }
                },
                anonymous: {
                  invoke: {
                    src: "signInAnonymous",
                    id: "authenticateAnonymously",
                    onDone: {
                      actions: ["saveSession", "reportTokenChanged"],
                      target: "#nhost.authentication.signedIn"
                    },
                    onError: {
                      actions: "saveAuthenticationError",
                      target: "#nhost.authentication.signedOut.failed"
                    }
                  }
                },
                mfa: {
                  states: {
                    totp: {
                      invoke: {
                        src: "signInMfaTotp",
                        id: "signInMfaTotp",
                        onDone: {
                          actions: ["saveSession", "reportTokenChanged"],
                          target: "#nhost.authentication.signedIn"
                        },
                        onError: {
                          actions: ["saveAuthenticationError"],
                          target: "#nhost.authentication.signedOut.failed"
                        }
                      }
                    }
                  }
                },
                securityKeyEmail: {
                  invoke: {
                    src: "signInSecurityKeyEmail",
                    id: "authenticateUserWithSecurityKey",
                    onDone: {
                      actions: ["saveSession", "reportTokenChanged"],
                      target: "#nhost.authentication.signedIn"
                    },
                    onError: [
                      {
                        cond: "unverified",
                        target: [
                          "#nhost.authentication.signedOut",
                          "#nhost.registration.incomplete.needsEmailVerification"
                        ]
                      },
                      {
                        actions: "saveAuthenticationError",
                        target: "#nhost.authentication.signedOut.failed"
                      }
                    ]
                  }
                }
              }
            },
            signedIn: {
              type: "parallel",
              entry: ["reportSignedIn", "cleanUrl", "broadcastToken", "resetErrors"],
              on: {
                SIGNOUT: "signedOut.signingOut"
              },
              states: {
                refreshTimer: {
                  id: "timer",
                  initial: "idle",
                  states: {
                    disabled: { type: "final" },
                    stopped: {
                      always: {
                        cond: "noToken",
                        target: "idle"
                      }
                    },
                    idle: {
                      always: [
                        { cond: "isAutoRefreshDisabled", target: "disabled" },
                        { cond: "isRefreshTokenPAT", target: "disabled" },
                        { cond: "hasRefreshToken", target: "running" }
                      ]
                    },
                    running: {
                      initial: "pending",
                      entry: "resetTimer",
                      states: {
                        pending: {
                          after: {
                            1e3: {
                              internal: !1,
                              target: "pending"
                            }
                          },
                          always: {
                            cond: "refreshTimerShouldRefresh",
                            target: "refreshing"
                          }
                        },
                        refreshing: {
                          invoke: {
                            src: "refreshToken",
                            id: "refreshToken",
                            onDone: {
                              actions: ["saveSession", "resetTimer", "reportTokenChanged"],
                              target: "pending"
                            },
                            onError: [{ actions: "saveRefreshAttempt", target: "pending" }]
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        },
        token: {
          initial: "idle",
          states: {
            idle: {
              on: {
                TRY_TOKEN: "running"
              },
              initial: "noErrors",
              states: { noErrors: {}, error: {} }
            },
            running: {
              invoke: {
                src: "refreshToken",
                id: "authenticateWithToken",
                onDone: {
                  actions: ["saveSession", "reportTokenChanged"],
                  target: ["#nhost.authentication.signedIn", "idle.noErrors"]
                },
                onError: [
                  { cond: "isSignedIn", target: "idle.error" },
                  {
                    actions: "saveAuthenticationError",
                    target: ["#nhost.authentication.signedOut.failed", "idle.error"]
                  }
                ]
              }
            }
          }
        },
        registration: {
          initial: "incomplete",
          on: {
            SIGNED_IN: [{ cond: "isAnonymous", target: ".incomplete" }, ".complete"]
          },
          states: {
            incomplete: {
              on: {
                SIGNUP_EMAIL_PASSWORD: "emailPassword",
                SIGNUP_SECURITY_KEY: "securityKey",
                PASSWORDLESS_EMAIL: "passwordlessEmail",
                PASSWORDLESS_SMS: "passwordlessSms",
                PASSWORDLESS_SMS_OTP: "passwordlessSmsOtp"
              },
              initial: "noErrors",
              states: {
                noErrors: {},
                needsEmailVerification: {},
                needsOtp: {},
                failed: {}
              }
            },
            emailPassword: {
              entry: ["resetErrors"],
              invoke: {
                src: "signUpEmailPassword",
                id: "signUpEmailPassword",
                onDone: [
                  {
                    cond: "hasSession",
                    actions: ["saveSession", "reportTokenChanged"],
                    target: "#nhost.authentication.signedIn"
                  },
                  {
                    actions: "clearContext",
                    target: ["#nhost.authentication.signedOut", "incomplete.needsEmailVerification"]
                  }
                ],
                onError: [
                  {
                    cond: "unverified",
                    target: "incomplete.needsEmailVerification"
                  },
                  {
                    actions: "saveRegistrationError",
                    target: "incomplete.failed"
                  }
                ]
              }
            },
            securityKey: {
              entry: ["resetErrors"],
              invoke: {
                src: "signUpSecurityKey",
                id: "signUpSecurityKey",
                onDone: [
                  {
                    cond: "hasSession",
                    actions: ["saveSession", "reportTokenChanged"],
                    target: "#nhost.authentication.signedIn"
                  },
                  {
                    actions: "clearContext",
                    target: ["#nhost.authentication.signedOut", "incomplete.needsEmailVerification"]
                  }
                ],
                onError: [
                  {
                    cond: "unverified",
                    target: "incomplete.needsEmailVerification"
                  },
                  {
                    actions: "saveRegistrationError",
                    target: "incomplete.failed"
                  }
                ]
              }
            },
            passwordlessEmail: {
              entry: ["resetErrors"],
              invoke: {
                src: "passwordlessEmail",
                id: "passwordlessEmail",
                onDone: {
                  actions: "clearContext",
                  target: ["#nhost.authentication.signedOut", "incomplete.needsEmailVerification"]
                },
                onError: {
                  actions: "saveRegistrationError",
                  target: "incomplete.failed"
                }
              }
            },
            passwordlessSms: {
              entry: ["resetErrors"],
              invoke: {
                src: "passwordlessSms",
                id: "passwordlessSms",
                onDone: {
                  actions: "clearContext",
                  target: ["#nhost.authentication.signedOut", "incomplete.needsOtp"]
                },
                onError: {
                  actions: "saveRegistrationError",
                  target: "incomplete.failed"
                }
              }
            },
            passwordlessSmsOtp: {
              entry: ["resetErrors"],
              invoke: {
                src: "passwordlessSmsOtp",
                id: "passwordlessSmsOtp",
                onDone: {
                  actions: ["saveSession", "reportTokenChanged"],
                  target: "#nhost.authentication.signedIn"
                },
                onError: {
                  actions: "saveRegistrationError",
                  target: "incomplete.failed"
                }
              }
            },
            complete: {
              on: {
                SIGNED_OUT: "incomplete"
              }
            }
          }
        }
      }
    },
    {
      actions: {
        reportSignedIn: p("SIGNED_IN"),
        reportSignedOut: p("SIGNED_OUT"),
        reportTokenChanged: p("TOKEN_CHANGED"),
        incrementTokenImportAttempts: E({
          importTokenAttempts: ({ importTokenAttempts: a }) => a + 1
        }),
        clearContext: E(() => (f(R, null), f(O, null), f(N, null), {
          ...G
        })),
        clearContextExceptRefreshToken: E(({ refreshToken: { value: a } }) => (f(R, null), {
          ...G,
          refreshToken: { value: a }
        })),
        // * Save session in the context, and persist the refresh token and the jwt expiration outside of the machine
        saveSession: E({
          user: (a, { data: o }) => {
            var c;
            return ((c = o == null ? void 0 : o.session) == null ? void 0 : c.user) || null;
          },
          accessToken: (a, { data: o }) => {
            if (o.session) {
              const { accessTokenExpiresIn: c, accessToken: l } = o.session, m = new Date(Date.now() + c * 1e3);
              return f(R, m.toISOString()), {
                value: l,
                expiresAt: m,
                expiresInSeconds: c
              };
            }
            return f(R, null), {
              value: null,
              expiresAt: null,
              expiresInSeconds: null
            };
          },
          refreshToken: (a, { data: o }) => {
            var m, T;
            const c = ((m = o.session) == null ? void 0 : m.refreshToken) || null, l = ((T = o.session) == null ? void 0 : T.refreshTokenId) || null;
            return c && f(O, c), l && f(N, l), { value: c };
          }
        }),
        savePATSession: E({
          user: (a, { data: o }) => {
            var c;
            return ((c = o == null ? void 0 : o.session) == null ? void 0 : c.user) || null;
          },
          accessToken: (a, { data: o }) => {
            if (o.session) {
              const { accessTokenExpiresIn: c, accessToken: l } = o.session, m = new Date(Date.now() + c * 1e3);
              return f(R, m.toISOString()), {
                value: l,
                expiresAt: m,
                expiresInSeconds: c
              };
            }
            return f(R, null), {
              value: null,
              expiresAt: null,
              expiresInSeconds: null
            };
          },
          refreshToken: (a, { data: o }) => {
            var m, T;
            const c = ((m = o.session) == null ? void 0 : m.refreshToken) || null, l = ((T = o.session) == null ? void 0 : T.refreshTokenId) || null;
            return c && f(O, c), l && f(N, l), { value: c, isPAT: !0 };
          }
        }),
        saveMfaTicket: E({
          mfa: (a, o) => {
            var c;
            return (c = o.data) == null ? void 0 : c.mfa;
          }
        }),
        resetTimer: E({
          refreshTimer: (a) => ({
            startedAt: /* @__PURE__ */ new Date(),
            attempts: 0,
            lastAttempt: null
          })
        }),
        saveRefreshAttempt: E({
          refreshTimer: (a, o) => ({
            startedAt: a.refreshTimer.startedAt,
            attempts: a.refreshTimer.attempts + 1,
            lastAttempt: /* @__PURE__ */ new Date()
          })
        }),
        // * Authentication errors
        saveAuthenticationError: E({
          // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
          errors: ({ errors: a }, { data: { error: o } }) => ({
            ...a,
            authentication: o
          })
        }),
        resetErrors: E({
          errors: (a) => ({}),
          importTokenAttempts: (a) => 0
        }),
        saveRegistrationError: E({
          // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
          errors: ({ errors: a }, { data: { error: o } }) => ({ ...a, registration: o })
        }),
        destroyRefreshToken: E({
          refreshToken: (a) => (f(O, null), f(N, null), { value: null })
        }),
        // * Clean the browser url when `autoSignIn` is activated
        cleanUrl: () => {
          u && D("refreshToken") && (F("refreshToken"), F("type"));
        },
        // * Broadcast the token to other tabs when `autoSignIn` is activated
        broadcastToken: (a) => {
          if (u)
            try {
              new BroadcastChannel("nhost").postMessage(a.refreshToken.value);
            } catch {
            }
        }
      },
      guards: {
        isAnonymous: (a, o) => {
          var c;
          return !!((c = a.user) != null && c.isAnonymous);
        },
        isSignedIn: (a) => !!a.user && !!a.refreshToken.value && !!a.accessToken.value,
        noToken: (a) => !a.refreshToken.value,
        isRefreshTokenPAT: (a) => {
          var o;
          return !!((o = a.refreshToken) != null && o.isPAT);
        },
        hasRefreshToken: (a) => !!a.refreshToken.value,
        isAutoRefreshDisabled: () => !i,
        refreshTimerShouldRefresh: (a) => {
          const { expiresAt: o } = a.accessToken;
          if (!o)
            return !1;
          if (a.refreshTimer.lastAttempt)
            return a.refreshTimer.attempts > H ? !1 : Date.now() - a.refreshTimer.lastAttempt.getTime() > Math.pow(2, a.refreshTimer.attempts - 1) * 5e3;
          if (s && Date.now() - a.refreshTimer.startedAt.getTime() > s * 1e3)
            return !0;
          const c = a.accessToken.expiresInSeconds;
          return c ? o.getTime() - Date.now() - 1e3 * Math.min(de, c * 0.5) <= 0 : !1;
        },
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        /** Shoud retry to import the token on network error or any internal server error.
         * Don't retry more than REFRESH_TOKEN_MAX_ATTEMPTS times.
         */
        shouldRetryImportToken: (a, o) => a.importTokenAttempts < H && (o.data.error.status === Z || o.data.error.status >= 500),
        // * Authentication errors
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        unverified: (a, { data: { error: o } }) => o.status === 401 && // * legacy: don't use the message contents to determine if the email is unverified, but the error type (error.error)
        (o.message === "Email is not verified" || o.error === "unverified-user"),
        // * Event guards
        hasSession: (a, o) => {
          var c;
          return !!((c = o.data) != null && c.session);
        },
        hasMfaTicket: (a, o) => {
          var c;
          return !!((c = o.data) != null && c.mfa);
        }
      },
      services: {
        signInPassword: (a, { email: o, password: c }) => v(o) ? W(c) ? h("/signin/email-password", {
          email: o,
          password: c
        }) : Promise.reject({ error: $ }) : Promise.reject({ error: I }),
        signInPAT: (a, { pat: o }) => h("/signin/pat", {
          personalAccessToken: o
        }),
        passwordlessSms: (a, { phoneNumber: o, options: c }) => {
          var l;
          return z(o) ? (l = a.user) != null && l.isAnonymous ? (console.warn(
            "Deanonymisation from a phone number is not yet implemented in hasura-auth"
          ), h(
            "/user/deanonymize",
            {
              signInMethod: "passwordless",
              connection: "sms",
              phoneNumber: o,
              options: w(e, c)
            },
            a.accessToken.value
          )) : h("/signin/passwordless/sms", {
            phoneNumber: o,
            options: w(e, c)
          }) : Promise.reject({ error: Y });
        },
        passwordlessSmsOtp: (a, { phoneNumber: o, otp: c }) => z(o) ? h("/signin/passwordless/sms/otp", {
          phoneNumber: o,
          otp: c
        }) : Promise.reject({ error: Y }),
        passwordlessEmail: (a, { email: o, options: c }) => {
          var l;
          return v(o) ? (l = a.user) != null && l.isAnonymous ? h(
            "/user/deanonymize",
            {
              signInMethod: "passwordless",
              connection: "email",
              email: o,
              options: w(e, c)
            },
            a.accessToken.value
          ) : h("/signin/passwordless/email", {
            email: o,
            options: w(e, c)
          }) : Promise.reject({ error: I });
        },
        signInAnonymous: (a) => h("/signin/anonymous"),
        signInMfaTotp: (a, o) => {
          var l;
          const c = o.ticket || ((l = a.mfa) == null ? void 0 : l.ticket);
          return c ? Ue(c) ? h("/signin/mfa/totp", {
            ticket: c,
            otp: o.otp
          }) : Promise.reject({ error: me }) : Promise.reject({ error: Ee });
        },
        signInSecurityKeyEmail: async (a, { email: o }) => {
          if (!v(o))
            throw new U(I);
          const c = await h(
            "/signin/webauthn",
            { email: o }
          );
          let l;
          try {
            l = await Pe(c);
          } catch (m) {
            throw new U(m);
          }
          return h("/signin/webauthn/verify", { email: o, credential: l });
        },
        refreshToken: async (a, o) => {
          const c = o.type === "TRY_TOKEN" ? o.token : a.refreshToken.value;
          return { session: await h("/token", {
            refreshToken: c
          }), error: null };
        },
        signout: (a, o) => h("/signout", {
          refreshToken: a.refreshToken.value,
          all: !!o.all
        }),
        signUpEmailPassword: async (a, { email: o, password: c, options: l }) => {
          var m;
          return v(o) ? W(c) ? (m = a.user) != null && m.isAnonymous ? h(
            "/user/deanonymize",
            {
              signInMethod: "email-password",
              email: o,
              password: c,
              options: w(e, l)
            },
            a.accessToken.value
          ) : h("/signup/email-password", {
            email: o,
            password: c,
            options: w(e, l)
          }) : Promise.reject({ error: $ }) : Promise.reject({ error: I });
        },
        signUpSecurityKey: async (a, { email: o, options: c }) => {
          if (!v(o))
            return Promise.reject({ error: I });
          const l = c == null ? void 0 : c.nickname;
          l && delete c.nickname;
          const m = await h(
            "/signup/webauthn",
            { email: o, options: c }
          );
          let T;
          try {
            T = await ne(m);
          } catch (ae) {
            throw new U(ae);
          }
          return h("/signup/webauthn/verify", {
            credential: T,
            options: {
              redirectTo: c == null ? void 0 : c.redirectTo,
              nickname: l
            }
          });
        },
        importRefreshToken: async (a) => {
          if (a.user && a.refreshToken.value && a.accessToken.value && a.accessToken.expiresAt)
            return {
              session: {
                accessToken: a.accessToken.value,
                accessTokenExpiresIn: a.accessToken.expiresAt.getTime() - Date.now(),
                refreshToken: a.refreshToken.value,
                user: a.user
              },
              error: null
            };
          let o = null;
          if (u) {
            const l = D("refreshToken") || null;
            if (l)
              try {
                return { session: await h("/token", {
                  refreshToken: l
                }), error: null };
              } catch (m) {
                o = m.error;
              }
            else {
              const m = D("error");
              if (m)
                return Promise.reject({
                  session: null,
                  error: {
                    status: y,
                    error: m,
                    message: D("errorDescription") || m
                  }
                });
            }
          }
          const c = await d(O);
          if (c)
            try {
              return { session: await h("/token", {
                refreshToken: c
              }), error: null };
            } catch (l) {
              o = l.error;
            }
          return o ? Promise.reject({ error: o, session: null }) : { error: null, session: null };
        }
      },
      delays: {
        RETRY_IMPORT_TOKEN_DELAY: ({ importTokenAttempts: a }) => Math.pow(2, a - 1) * 5e3
      }
    }
  );
}, Ke = ({ backendUrl: n, clientUrl: e, interpreter: r }) => P(
  {
    schema: {
      context: {},
      events: {},
      services: {}
    },
    tsTypes: {},
    predictableActionArguments: !0,
    id: "changeEmail",
    initial: "idle",
    context: { error: null },
    states: {
      idle: {
        on: {
          REQUEST: [
            {
              cond: "invalidEmail",
              actions: "saveInvalidEmailError",
              target: ".error"
            },
            {
              target: "requesting"
            }
          ]
        },
        initial: "initial",
        states: {
          initial: {},
          success: {},
          error: {}
        }
      },
      requesting: {
        invoke: {
          src: "requestChange",
          id: "requestChange",
          onDone: { target: "idle.success", actions: "reportSuccess" },
          onError: { actions: ["saveRequestError", "reportError"], target: "idle.error" }
        }
      }
    }
  },
  {
    actions: {
      saveInvalidEmailError: E({ error: (t) => I }),
      saveRequestError: E({
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        error: (t, { data: { error: s } }) => s
      }),
      reportError: p((t) => ({ type: "ERROR", error: t.error })),
      // TODO change email in the main machine (context.user.email)
      reportSuccess: p("SUCCESS")
    },
    guards: {
      invalidEmail: (t, { email: s }) => !v(s)
    },
    services: {
      requestChange: async (t, { email: s, options: i }) => (await k(
        `${n}/user/email/change`,
        { newEmail: s, options: w(e, i) },
        r == null ? void 0 : r.getSnapshot().context.accessToken.value
      )).data
    }
  }
), Ve = ({ backendUrl: n, interpreter: e }) => P(
  {
    schema: {
      context: {},
      events: {},
      services: {}
    },
    tsTypes: {},
    predictableActionArguments: !0,
    id: "changePassword",
    initial: "idle",
    context: { error: null },
    states: {
      idle: {
        on: {
          REQUEST: [
            {
              cond: "invalidPassword",
              actions: "saveInvalidPasswordError",
              target: ".error"
            },
            {
              target: "requesting"
            }
          ]
        },
        initial: "initial",
        states: {
          initial: {},
          success: {},
          error: {}
        }
      },
      requesting: {
        invoke: {
          src: "requestChange",
          id: "requestChange",
          onDone: { target: "idle.success", actions: "reportSuccess" },
          onError: { actions: ["saveRequestError", "reportError"], target: "idle.error" }
        }
      }
    }
  },
  {
    actions: {
      saveInvalidPasswordError: E({ error: (r) => $ }),
      saveRequestError: E({
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        error: (r, { data: { error: t } }) => t
      }),
      reportError: p((r) => ({ type: "ERROR", error: r.error })),
      reportSuccess: p("SUCCESS")
    },
    guards: {
      invalidPassword: (r, { password: t }) => !W(t)
    },
    services: {
      requestChange: (r, { password: t, ticket: s }) => k(
        `${n}/user/password`,
        { newPassword: t, ticket: s },
        e == null ? void 0 : e.getSnapshot().context.accessToken.value
      )
    }
  }
), ar = ({ backendUrl: n, interpreter: e }) => P(
  {
    schema: {
      context: {},
      events: {}
    },
    tsTypes: {},
    predictableActionArguments: !0,
    id: "enableMfa",
    initial: "idle",
    context: { error: null, imageUrl: null, secret: null },
    states: {
      idle: {
        initial: "initial",
        on: {
          GENERATE: "generating"
        },
        states: {
          initial: {},
          error: {}
        }
      },
      generating: {
        invoke: {
          src: "generate",
          id: "generate",
          onDone: { target: "generated", actions: ["reportGeneratedSuccess", "saveGeneration"] },
          onError: { actions: ["saveError", "reportGeneratedError"], target: "idle.error" }
        }
      },
      generated: {
        initial: "idle",
        states: {
          idle: {
            initial: "idle",
            on: {
              ACTIVATE: [
                {
                  cond: "invalidMfaType",
                  actions: "saveInvalidMfaTypeError",
                  target: ".error"
                },
                {
                  cond: "invalidMfaCode",
                  actions: "saveInvalidMfaCodeError",
                  target: ".error"
                },
                {
                  target: "activating"
                }
              ]
            },
            states: { idle: {}, error: {} }
          },
          activating: {
            invoke: {
              src: "activate",
              id: "activate",
              onDone: { target: "activated", actions: "reportSuccess" },
              onError: { actions: ["saveError", "reportError"], target: "idle.error" }
            }
          },
          activated: { type: "final" }
        }
      }
    }
  },
  {
    actions: {
      saveInvalidMfaTypeError: E({ error: (r) => fe }),
      saveInvalidMfaCodeError: E({ error: (r) => he }),
      saveError: E({
        error: (r, { data: { error: t } }) => t
      }),
      saveGeneration: E({
        imageUrl: (r, { data: { imageUrl: t } }) => t,
        secret: (r, { data: { totpSecret: t } }) => t
      }),
      reportError: p((r, t) => (console.log("REPORT", r, t), { type: "ERROR", error: r.error })),
      reportSuccess: p("SUCCESS"),
      reportGeneratedSuccess: p("GENERATED"),
      reportGeneratedError: p((r) => ({ type: "GENERATED_ERROR", error: r.error }))
    },
    guards: {
      invalidMfaCode: (r, { code: t }) => !t,
      invalidMfaType: (r, { activeMfaType: t }) => !t || t !== "totp"
    },
    services: {
      generate: async (r) => {
        const { data: t } = await De(
          `${n}/mfa/totp/generate`,
          e == null ? void 0 : e.getSnapshot().context.accessToken.value
        );
        return t;
      },
      activate: (r, { code: t, activeMfaType: s }) => k(
        `${n}/user/mfa`,
        { code: t, activeMfaType: s },
        e == null ? void 0 : e.getSnapshot().context.accessToken.value
      )
    }
  }
), Ge = ({ backendUrl: n, clientUrl: e }) => P(
  {
    schema: {
      context: {},
      events: {},
      services: {}
    },
    tsTypes: {},
    predictableActionArguments: !0,
    id: "changePassword",
    initial: "idle",
    context: { error: null },
    states: {
      idle: {
        on: {
          REQUEST: [
            {
              cond: "invalidEmail",
              actions: "saveInvalidEmailError",
              target: ".error"
            },
            {
              target: "requesting"
            }
          ]
        },
        initial: "initial",
        states: {
          initial: {},
          success: {},
          error: {}
        }
      },
      requesting: {
        invoke: {
          src: "requestChange",
          id: "requestChange",
          onDone: { target: "idle.success", actions: "reportSuccess" },
          onError: { actions: ["saveRequestError", "reportError"], target: "idle.error" }
        }
      }
    }
  },
  {
    actions: {
      saveInvalidEmailError: E({ error: (r) => I }),
      saveRequestError: E({
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        error: (r, { data: { error: t } }) => t
      }),
      reportError: p((r) => ({ type: "ERROR", error: r.error })),
      reportSuccess: p("SUCCESS")
    },
    guards: {
      invalidEmail: (r, { email: t }) => !v(t)
    },
    services: {
      requestChange: (r, { email: t, options: s }) => k(`${n}/user/password/reset`, {
        email: t,
        options: w(e, s)
      })
    }
  }
), Le = ({ backendUrl: n, clientUrl: e }) => P(
  {
    schema: {
      context: {},
      events: {},
      services: {}
    },
    tsTypes: {},
    predictableActionArguments: !0,
    id: "sendVerificationEmail",
    initial: "idle",
    context: { error: null },
    states: {
      idle: {
        on: {
          REQUEST: [
            {
              cond: "invalidEmail",
              actions: "saveInvalidEmailError",
              target: ".error"
            },
            {
              target: "requesting"
            }
          ]
        },
        initial: "initial",
        states: {
          initial: {},
          success: {},
          error: {}
        }
      },
      requesting: {
        invoke: {
          src: "request",
          id: "request",
          onDone: { target: "idle.success", actions: "reportSuccess" },
          onError: { actions: ["saveRequestError", "reportError"], target: "idle.error" }
        }
      }
    }
  },
  {
    actions: {
      saveInvalidEmailError: E({ error: (r) => I }),
      saveRequestError: E({
        // * Untyped action payload. See https://github.com/statelyai/xstate/issues/3037
        error: (r, { data: { error: t } }) => t
      }),
      reportError: p((r) => ({ type: "ERROR", error: r.error })),
      reportSuccess: p("SUCCESS")
    },
    guards: {
      invalidEmail: (r, { email: t }) => !v(t)
    },
    services: {
      request: async (r, { email: t, options: s }) => (await k(
        `${n}/user/email/send-verification-email`,
        { email: t, options: w(e, s) }
      )).data
    }
  }
);
class oe {
  constructor({
    clientStorageType: e = "web",
    autoSignIn: r = !0,
    autoRefreshToken: t = !0,
    start: s = !0,
    backendUrl: i,
    clientUrl: u,
    devTools: d,
    ...f
  }) {
    if (this._started = !1, this._subscriptionsQueue = /* @__PURE__ */ new Set(), this._subscriptions = /* @__PURE__ */ new Set(), this.backendUrl = i, this.clientUrl = u, this._machine = Me({
      ...f,
      backendUrl: i,
      clientUrl: u,
      clientStorageType: e,
      autoSignIn: r,
      autoRefreshToken: t
    }), s && this.start({ devTools: d }), typeof window != "undefined" && r)
      try {
        this._channel = new BroadcastChannel("nhost"), this._channel.addEventListener("message", (h) => {
          var o;
          const a = (o = this.interpreter) == null ? void 0 : o.getSnapshot().context.refreshToken.value;
          this.interpreter && h.data !== a && this.interpreter.send("TRY_TOKEN", { token: h.data });
        });
      } catch {
      }
  }
  start({
    devTools: e = !1,
    initialSession: r,
    interpreter: t
  } = {}) {
    var u, d;
    const s = { ...this.machine.context };
    r && (s.user = r.user, s.refreshToken.value = (u = r.refreshToken) != null ? u : null, s.accessToken.value = (d = r.accessToken) != null ? d : null, s.accessToken.expiresAt = new Date(
      Date.now() + r.accessTokenExpiresIn * 1e3
    ));
    const i = this.machine.withContext(s);
    this._interpreter || (this._interpreter = t || b(i, { devTools: e })), (!this._started || typeof window == "undefined") && (this._interpreter.initialized && (this._interpreter.stop(), this._subscriptions.forEach((f) => f())), this._interpreter.start(i.initialState), this._subscriptionsQueue.forEach((f) => f(this))), this._started = !0;
  }
  get machine() {
    return this._machine;
  }
  get interpreter() {
    return this._interpreter;
  }
  get started() {
    return this._started;
  }
  subscribe(e) {
    if (this.started) {
      const r = e(this);
      return this._subscriptions.add(r), r;
    } else
      return this._subscriptionsQueue.add(e), () => {
        console.log(
          "onTokenChanged was added before the interpreter started. Cannot unsubscribe listener."
        );
      };
  }
}
class $e extends oe {
  constructor({
    ...e
  }) {
    super({
      ...e,
      autoSignIn: q() && e.autoSignIn,
      autoRefreshToken: q() && e.autoRefreshToken,
      clientStorageType: "cookie"
    });
  }
}
const cr = $e, qe = async ({ backendUrl: n, interpreter: e }, r) => {
  try {
    const { data: t } = await k(
      `${n}/user/webauthn/add`,
      {},
      e == null ? void 0 : e.getSnapshot().context.accessToken.value
    );
    let s;
    try {
      s = await ne(t);
    } catch (u) {
      throw new U(u);
    }
    const { data: i } = await k(
      `${n}/user/webauthn/verify`,
      { credential: s, nickname: r },
      e == null ? void 0 : e.getSnapshot().context.accessToken.value
    );
    return { key: i, isError: !1, error: null, isSuccess: !0 };
  } catch (t) {
    const { error: s } = t;
    return { isError: !0, error: s, isSuccess: !1 };
  }
}, We = async (n, e, r) => new Promise((t) => {
  n.send("REQUEST", {
    email: e,
    options: r
  }), n.onTransition((s) => {
    s.matches({ idle: "error" }) ? t({ error: s.context.error, isError: !0, needsEmailVerification: !1 }) : s.matches({ idle: "success" }) && t({ error: null, isError: !1, needsEmailVerification: !0 });
  });
}), je = async (n, e, r) => new Promise((t) => {
  n.send("REQUEST", {
    password: e,
    ticket: r
  }), n.onTransition((s) => {
    s.matches({ idle: "error" }) ? t({ error: s.context.error, isError: !0, isSuccess: !1 }) : s.matches({ idle: "success" }) && t({ error: null, isError: !1, isSuccess: !0 });
  });
}), ur = (n) => new Promise((e) => {
  n.send("GENERATE"), n.onTransition((r) => {
    r.matches("generated") ? e({
      error: null,
      isError: !1,
      isGenerated: !0,
      qrCodeDataUrl: r.context.imageUrl || ""
    }) : r.matches({ idle: "error" }) && e({
      error: r.context.error || null,
      isError: !0,
      isGenerated: !1,
      qrCodeDataUrl: ""
    });
  });
}), lr = (n, e) => new Promise((r) => {
  n.send("ACTIVATE", {
    activeMfaType: "totp",
    code: e
  }), n.onTransition((t) => {
    t.matches({ generated: "activated" }) ? r({ error: null, isActivated: !0, isError: !1 }) : t.matches({ generated: { idle: "error" } }) && r({ error: t.context.error, isActivated: !1, isError: !0 });
  });
}), He = async (n, e, r) => new Promise((t) => {
  n.send("REQUEST", {
    email: e,
    options: r
  }), n.onTransition((s) => {
    s.matches({ idle: "error" }) ? t({ error: s.context.error, isError: !0, isSent: !1 }) : s.matches({ idle: "success" }) && t({ error: null, isError: !1, isSent: !0 });
  });
}), Ye = (n, e, r) => new Promise((t) => {
  n.send("REQUEST", {
    email: e,
    options: r
  }), n.onTransition((s) => {
    s.matches({ idle: "error" }) ? t({ error: s.context.error, isError: !0, isSent: !1 }) : s.matches({ idle: "success" }) && t({ error: null, isError: !1, isSent: !0 });
  });
}), Fe = (n) => new Promise((e) => {
  const { changed: r } = n.send("SIGNIN_ANONYMOUS");
  r || e({
    isSuccess: !1,
    isError: !0,
    error: S,
    user: null,
    accessToken: null
  }), n.onTransition((t) => {
    t.matches({ authentication: "signedIn" }) && e({
      isSuccess: !0,
      isError: !1,
      error: null,
      user: t.context.user,
      accessToken: t.context.accessToken.value
    }), t.matches({ authentication: { signedOut: "failed" } }) && e({
      isSuccess: !1,
      isError: !0,
      error: t.context.errors.authentication || null,
      user: null,
      accessToken: null
    });
  });
}), ze = (n, e, r) => new Promise((t) => {
  const { changed: s, context: i } = n.send("SIGNIN_PASSWORD", {
    email: e,
    password: r
  });
  if (!s)
    return t({
      accessToken: i.accessToken.value,
      error: S,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      needsMfaOtp: !1,
      mfa: null,
      user: i.user
    });
  n.onTransition((u) => {
    u.matches({
      authentication: { signedOut: "noErrors" },
      registration: { incomplete: "needsEmailVerification" }
    }) ? t({
      accessToken: null,
      error: null,
      isError: !1,
      isSuccess: !1,
      needsEmailVerification: !0,
      needsMfaOtp: !1,
      mfa: null,
      user: null
    }) : u.matches({ authentication: { signedOut: "needsMfa" } }) ? t({
      accessToken: null,
      error: null,
      isError: !1,
      isSuccess: !1,
      needsEmailVerification: !1,
      needsMfaOtp: !0,
      mfa: u.context.mfa,
      user: null
    }) : u.matches({ authentication: { signedOut: "failed" } }) ? t({
      accessToken: null,
      error: u.context.errors.authentication || null,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      needsMfaOtp: !1,
      mfa: null,
      user: null
    }) : u.matches({ authentication: "signedIn" }) && t({
      accessToken: u.context.accessToken.value,
      error: null,
      isError: !1,
      isSuccess: !0,
      needsEmailVerification: !1,
      needsMfaOtp: !1,
      mfa: null,
      user: u.context.user
    });
  });
}), Q = (n, e, r) => new Promise((t) => {
  const { changed: s } = n.send("PASSWORDLESS_EMAIL", {
    email: e,
    options: r
  });
  if (!s)
    return t({
      error: S,
      isError: !0,
      isSuccess: !1
    });
  n.onTransition((i) => {
    i.matches("registration.incomplete.failed") ? t({
      error: i.context.errors.registration || null,
      isError: !0,
      isSuccess: !1
    }) : i.matches({
      authentication: { signedOut: "noErrors" },
      registration: { incomplete: "needsEmailVerification" }
    }) && t({ error: null, isError: !1, isSuccess: !0 });
  });
}), Qe = (n, e) => new Promise((r) => {
  const { changed: t, context: s } = n.send({ type: "SIGNIN_SECURITY_KEY_EMAIL", email: e });
  if (!t)
    return r({
      accessToken: s.accessToken.value,
      error: S,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: s.user
    });
  n.onTransition((i) => {
    i.matches({
      authentication: { signedOut: "noErrors" },
      registration: { incomplete: "needsEmailVerification" }
    }) ? r({
      accessToken: null,
      error: null,
      isError: !1,
      isSuccess: !1,
      needsEmailVerification: !0,
      user: null
    }) : i.matches({ authentication: { signedOut: "failed" } }) ? r({
      accessToken: null,
      error: i.context.errors.authentication || null,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: null
    }) : i.matches({ authentication: "signedIn" }) && r({
      accessToken: i.context.accessToken.value,
      error: null,
      isError: !1,
      isSuccess: !0,
      needsEmailVerification: !1,
      user: i.context.user
    });
  });
}), Be = (n, e, r) => new Promise((t) => {
  const { changed: s, context: i } = n.send("SIGNIN_MFA_TOTP", {
    otp: e,
    ticket: r
  });
  if (!s)
    return t({
      accessToken: i.accessToken.value,
      error: S,
      isError: !0,
      isSuccess: !1,
      user: i.user
    });
  n.onTransition((u) => {
    u.matches({ authentication: { signedOut: "failed" } }) ? t({
      accessToken: null,
      error: u.context.errors.authentication || null,
      isError: !0,
      isSuccess: !1,
      user: null
    }) : u.matches({ authentication: "signedIn" }) && t({
      accessToken: u.context.accessToken.value,
      error: null,
      isError: !1,
      isSuccess: !0,
      user: u.context.user
    });
  });
}), Xe = (n, e) => new Promise((r) => {
  const { changed: t } = n.send("SIGNIN_PAT", { pat: e });
  t || r({
    isSuccess: !1,
    isError: !0,
    error: S,
    user: null,
    accessToken: null
  }), n.onTransition((s) => {
    if (s.matches({ authentication: { signedOut: "failed" } }))
      return r({
        accessToken: null,
        user: null,
        error: s.context.errors.authentication || null,
        isError: !0,
        isSuccess: !1
      });
    if (s.matches({ authentication: "signedIn" }))
      return r({
        accessToken: s.context.accessToken.value,
        user: s.context.user,
        error: null,
        isError: !1,
        isSuccess: !0
      });
  });
}), B = (n, e, r) => new Promise((t) => {
  const { changed: s } = n.send("PASSWORDLESS_SMS", { phoneNumber: e, options: r });
  if (!s)
    return t({
      error: S,
      isError: !0,
      isSuccess: !1,
      needsOtp: !1
    });
  n.onTransition((i) => {
    i.matches("registration.incomplete.needsOtp") ? t({
      error: null,
      isError: !1,
      isSuccess: !1,
      needsOtp: !0
    }) : i.matches("registration.incomplete.failed") && t({
      error: i.context.errors.authentication || null,
      isError: !0,
      isSuccess: !1,
      needsOtp: !1
    });
  });
}), Ze = (n, e, r) => new Promise((t) => {
  const { changed: s } = n.send({ type: "PASSWORDLESS_SMS_OTP", phoneNumber: e, otp: r });
  if (!s)
    return t({
      error: S,
      isError: !0,
      isSuccess: !1,
      user: null,
      accessToken: null
    });
  n.onTransition((i) => {
    i.matches({ authentication: "signedIn" }) ? t({
      error: null,
      isError: !1,
      isSuccess: !0,
      user: i.context.user,
      accessToken: i.context.accessToken.value
    }) : i.matches({ registration: { incomplete: "failed" } }) && t({
      error: i.context.errors.authentication || null,
      isError: !0,
      isSuccess: !1,
      user: null,
      accessToken: null
    });
  });
}), Je = async (n, e) => new Promise((r) => {
  const { event: t } = n.send("SIGNOUT", { all: e });
  if (t.type !== "SIGNED_OUT")
    return r({ isSuccess: !1, isError: !0, error: Te });
  n.onTransition((s) => {
    s.matches({ authentication: { signedOut: "success" } }) ? r({ isSuccess: !0, isError: !1, error: null }) : s.matches("authentication.signedOut.failed") && r({ isSuccess: !1, isError: !0, error: s.context.errors.signout || null });
  });
}), X = (n, e, r, t) => new Promise((s) => {
  const { changed: i, context: u } = n.send("SIGNUP_EMAIL_PASSWORD", {
    email: e,
    password: r,
    options: t
  });
  if (!i)
    return s({
      error: S,
      accessToken: u.accessToken.value,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: u.user
    });
  n.onTransition((d) => {
    d.matches("registration.incomplete.failed") ? s({
      accessToken: null,
      error: d.context.errors.registration || null,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: null
    }) : d.matches({
      authentication: { signedOut: "noErrors" },
      registration: { incomplete: "needsEmailVerification" }
    }) ? s({
      accessToken: null,
      error: null,
      isError: !1,
      isSuccess: !1,
      needsEmailVerification: !0,
      user: null
    }) : d.matches({ authentication: "signedIn", registration: "complete" }) && s({
      accessToken: d.context.accessToken.value,
      error: null,
      isError: !1,
      isSuccess: !0,
      needsEmailVerification: !1,
      user: d.context.user
    });
  });
}), er = (n, e, r) => new Promise((t) => {
  const { changed: s, context: i } = n.send("SIGNUP_SECURITY_KEY", {
    email: e,
    options: r
  });
  if (!s)
    return t({
      error: S,
      accessToken: i.accessToken.value,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: i.user
    });
  n.onTransition((u) => {
    u.matches("registration.incomplete.failed") ? t({
      accessToken: null,
      error: u.context.errors.registration || null,
      isError: !0,
      isSuccess: !1,
      needsEmailVerification: !1,
      user: null
    }) : u.matches({
      authentication: { signedOut: "noErrors" },
      registration: { incomplete: "needsEmailVerification" }
    }) ? t({
      accessToken: null,
      error: null,
      isError: !1,
      isSuccess: !1,
      needsEmailVerification: !0,
      user: null
    }) : u.matches({ authentication: "signedIn", registration: "complete" }) && t({
      accessToken: u.context.accessToken.value,
      error: null,
      isError: !1,
      isSuccess: !0,
      needsEmailVerification: !1,
      user: u.context.user
    });
  });
}), rr = async ({ backendUrl: n, interpreter: e }, { expiresAt: r, metadata: t }) => {
  try {
    const { data: s } = await k(
      `${n}/pat`,
      { expiresAt: r.toUTCString(), metadata: t },
      e == null ? void 0 : e.getSnapshot().context.accessToken.value
    );
    return {
      data: s ? {
        id: s.id || null,
        personalAccessToken: s.personalAccessToken || null
      } : null,
      isError: !1,
      error: null,
      isSuccess: !0
    };
  } catch (s) {
    const { error: i } = s;
    return { isError: !0, error: i, isSuccess: !1, data: null };
  }
};
class dr {
  constructor({
    url: e,
    autoRefreshToken: r = !0,
    autoSignIn: t = !0,
    clientStorage: s,
    clientStorageType: i,
    refreshIntervalTime: u,
    start: d = !0
  }) {
    var f;
    this.url = e, this._client = new oe({
      backendUrl: e,
      clientUrl: typeof window != "undefined" && ((f = window.location) == null ? void 0 : f.origin) || "",
      autoRefreshToken: r,
      autoSignIn: t,
      start: d,
      clientStorage: s,
      clientStorageType: i,
      refreshIntervalTime: u
    });
  }
  /**
   * Use `nhost.auth.signUp` to sign up a user using email and password. If you want to sign up a user using passwordless email (Magic Link), SMS, or an OAuth provider, use the `signIn` function instead.
   *
   * @example
   * ### Sign up with an email and password
   * ```ts
   * nhost.auth.signUp({
   *   email: 'joe@example.com',
   *   password: 'secret-password'
   * })
   * ```
   *
   * @example
   * ### Sign up with a security key
   * ```ts
   * nhost.auth.signUp({
   *   email: 'joe@example.com',
   *   securityKey: true
   * })
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/sign-up
   */
  async signUp(e) {
    const r = await this.waitUntilReady(), { email: t, options: s } = e;
    return "securityKey" in e ? A(
      await er(r, t, s)
    ) : A(
      await X(r, t, e.password, s)
    );
  }
  /**
   * Use `nhost.auth.signIn` to sign in a user using email and password, passwordless (email or sms) or an external provider. `signIn` can be used to sign in a user in various ways depending on the parameters.
   *
   * @example
   * ### Sign in a user using email and password
   * ```ts
   * nhost.auth.signIn({
   *   email: 'joe@example.com',
   *   password: 'secret-password'
   * })
   * ```
   *
   * @example
   * ### Sign in a user using an OAuth provider (e.g: Google or Facebook)
   * ```ts
   * nhost.auth.signIn({ provider: 'google' })
   * ```
   *
   * @example
   * ### Sign in a user using passwordless email (Magic Link)
   * ```ts
   * nhost.auth.signIn({ email: 'joe@example.com' })
   * ```
   *
   * @example
   * ### Sign in a user using passwordless SMS
   * ```ts
   * // [step 1/2] Passwordless sign in using SMS
   * nhost.auth.signIn({ phoneNumber: '+11233213123' })
   *
   * // [step 2/2] Finish passwordless sign in using SMS (OTP)
   * nhost.auth.signIn({ phoneNumber: '+11233213123', otp: '123456' })
   * ```
   *
   * @example
   * ### Sign in anonymously
   * ```ts
   * // Sign in anonymously
   * nhost.auth.signIn()
   *
   * // Later in the application, the user can complete their registration
   * nhost.auth.signUp({
   *   email: 'joe@example.com',
   *   password: 'secret-password'
   * })
   * ```
   *
   * @example
   * ### Sign in with a security key
   * ```ts
   * nhost.auth.signIn({
   *   email: 'joe@example.com',
   *   securityKey: true
   * })
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/sign-in
   */
  async signIn(e) {
    const r = await this.waitUntilReady();
    if (!e) {
      const t = await Fe(r);
      return { ...A(t), mfa: null };
    }
    if ("provider" in e) {
      const { provider: t, options: s } = e, i = ie(
        `${this._client.backendUrl}/signin/provider/${t}`,
        w(this._client.clientUrl, s)
      );
      return q() && (window.location.href = i), { providerUrl: i, provider: t, session: null, mfa: null, error: null };
    }
    if ("email" in e && "password" in e) {
      const t = await ze(r, e.email, e.password);
      return t.needsEmailVerification ? { session: null, mfa: null, error: we } : t.needsMfaOtp ? {
        session: null,
        mfa: t.mfa,
        error: null
      } : { ...A(t), mfa: null };
    }
    if ("email" in e && "securityKey" in e) {
      if (e.securityKey !== !0)
        throw Error("securityKey must be true");
      const t = await Qe(r, e.email);
      return { ...A(t), mfa: null };
    }
    if ("email" in e) {
      const { email: t, options: s } = e, { error: i } = await Q(r, t, s);
      return {
        session: null,
        mfa: null,
        error: i
      };
    }
    if ("phoneNumber" in e && "otp" in e) {
      const t = await Ze(r, e.phoneNumber, e.otp);
      return { ...A(t), mfa: null };
    }
    if ("phoneNumber" in e) {
      const { error: t } = await B(
        r,
        e.phoneNumber,
        e.options
      );
      return { error: t, mfa: null, session: null };
    }
    if ("otp" in e) {
      const t = await Be(r, e.otp, e.ticket);
      return { ...A(t), mfa: null };
    }
    return { error: Se, mfa: null, session: null };
  }
  /**
   * Use `nhost.auth.signInPAT` to sign in with a personal access token (PAT).
   *
   * @example
   * ```ts
   * nhost.auth.signInPAT('34f74930-09c0-4af5-a8d5-28fad78e3415')
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/sign-in-pat
   *
   * @param personalAccessToken - The personal access token to sign in with
   */
  async signInPAT(e) {
    const r = await this.waitUntilReady(), t = await Xe(r, e);
    return A(t);
  }
  /**
   * Use `nhost.auth.signOut` to sign out the user.
   *
   * @example
   * ### Sign out the user from current device
   * ```ts
   * nhost.auth.signOut()
   * ```
   *
   * @example
   * ### Sign out the user from all devices
   * ```ts
   * nhost.auth.signOut({all: true})
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/sign-out
   */
  async signOut(e) {
    const r = await this.waitUntilReady(), { error: t } = await Je(r, e == null ? void 0 : e.all);
    return { error: t };
  }
  /**
   * Use `nhost.auth.resetPassword` to reset the password for a user. This will send a reset-password link in an email to the user. When the user clicks the reset-password link the user is automatically signed-in. Once signed-in, the user can change their password using `nhost.auth.changePassword()`.
   *
   * @example
   * ```ts
   * nhost.auth.resetPassword({email: 'joe@example.com' })
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/reset-password
   */
  async resetPassword({ email: e, options: r }) {
    const t = b(Ge(this._client)).start(), { error: s } = await He(t, e, r);
    return { error: s };
  }
  /**
   * Use `nhost.auth.changePassword` to change the password for the signed-in user. The old password is not needed. In case the user is not signed-in, a password reset ticket needs to be provided.
   *
   * @example
   * ```ts
   * nhost.auth.changePassword({ newPassword: 'new-secret-password' })
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/change-password
   */
  async changePassword({
    newPassword: e,
    ticket: r
  }) {
    const t = b(Ve(this._client)).start(), { error: s } = await je(t, e, r);
    return { error: s };
  }
  /**
   * Use `nhost.auth.sendVerificationEmail` to send a verification email to the specified email. The email contains a verification-email link. When the user clicks the verification-email link their email is verified.
   *
   * @example
   * ```ts
   * nhost.auth.sendVerificationEmail({ email: 'joe@example.com' })
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/send-verification-email
   */
  async sendVerificationEmail({
    email: e,
    options: r
  }) {
    const t = b(Le(this._client)).start(), { error: s } = await Ye(t, e, r);
    return { error: s };
  }
  /**
   * Use `nhost.auth.changeEmail` to change a user's email. This will send a confirm-email-change link in an email to the new email. Once the user clicks on the confirm-email-change link the email will be change to the new email.
   *
   * @example
   * ```ts
   * nhost.auth.changeEmail({ newEmail: 'doe@example.com' })
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/change-email
   */
  async changeEmail({ newEmail: e, options: r }) {
    const t = b(Ke(this._client)).start(), { error: s } = await We(t, e, r);
    return { error: s };
  }
  /**
   * Use `nhost.auth.deanonymize` to deanonymize a user.
   *
   * @example
   * ```ts
   * nhost.auth.deanonymize({signInMethod: 'email-password', email: 'joe@example.com' })
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/deanonymize
   */
  async deanonymize(e) {
    const r = await this.waitUntilReady();
    if (e.signInMethod === "passwordless") {
      if (e.connection === "email") {
        const { error: t } = await Q(
          r,
          e.email,
          e.options
        );
        return { error: t };
      }
      if (e.connection === "sms") {
        const { error: t } = await B(
          r,
          e.phoneNumber,
          e.options
        );
        return { error: t };
      }
    }
    if (e.signInMethod === "email-password") {
      const { error: t } = await X(
        r,
        e.email,
        e.password,
        e.options
      );
      return { error: t };
    }
    throw Error("Unknown deanonymization method");
  }
  /**
   * Use `nhost.auth.addSecurityKey to add a security key to the user, using the WebAuthn API.
   * @param nickname optional human-readable nickname for the security key
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/add-security-key
   */
  async addSecurityKey(e) {
    const { error: r, key: t } = await qe(this._client, e);
    return { error: r, key: t };
  }
  /**
   * Use `nhost.auth.createPAT` to create a personal access token for the user.
   *
   * @param expiresAt Expiration date for the token
   * @param metadata Optional metadata to store with the token
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/create-pat
   */
  async createPAT(e, r) {
    return rr(this._client, { expiresAt: e, metadata: r });
  }
  /**
   * Use `nhost.auth.onTokenChanged` to add a custom function that runs every time the access or refresh token is changed.
   *
   *
   * @example
   * ```ts
   * nhost.auth.onTokenChanged(() => console.log('The access and refresh token has changed'));
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/on-token-changed
   */
  onTokenChanged(e) {
    return this._client.subscribe(() => {
      var t;
      const r = (t = this._client.interpreter) == null ? void 0 : t.onTransition(({ event: s, context: i }) => {
        s.type === "TOKEN_CHANGED" && e(x(i));
      });
      return () => r == null ? void 0 : r.stop();
    });
  }
  /**
   * Use `nhost.auth.onAuthStateChanged` to add a custom function that runs every time the authentication status of the user changes. E.g. add a custom function that runs every time the authentication status changes from signed-in to signed-out.
   *
   * @example
   * ```ts
   * nhost.auth.onAuthStateChanged((event, session) => {
   *   console.log(`The auth state has changed. State is now ${event} with session: ${session}`)
   * });
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/on-auth-state-changed
   */
  onAuthStateChanged(e) {
    return this._client.subscribe(() => {
      var t;
      const r = (t = this._client.interpreter) == null ? void 0 : t.onTransition(({ event: s, context: i }) => {
        (s.type === "SIGNED_IN" || s.type === "SIGNED_OUT") && e(s.type, x(i));
      });
      return () => r == null ? void 0 : r.stop();
    });
  }
  /**
   * Use `nhost.auth.isAuthenticated` to check if the user is authenticated or not.
   *
   * Note: `nhost.auth.isAuthenticated()` can return `false` for two reasons:
   * 1. The user is not authenticated
   * 2. The user is not authenticated but _might_ be authenticated soon (loading) because there is a network request in transit.
   *
   * Use `nhost.auth.getAuthenticationStatus` to get both authentication and loading status.
   *
   * @example
   * ```ts
   * const isAuthenticated = nhost.auth.isAuthenticated();
   *
   * if (isAuthenticated) {
   *   console.log('User is authenticated');
   * }
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/is-authenticated
   */
  isAuthenticated() {
    var e;
    return !!((e = this._client.interpreter) != null && e.getSnapshot().matches({ authentication: "signedIn" }));
  }
  /**
   * Use `nhost.auth.isAuthenticatedAsync` to wait (await) for any internal authentication network requests to finish and then return the authentication status.
   *
   * The promise won't resolve until the authentication status is known.
   * Attention: when using auto-signin and a refresh token is present in the client storage, the promise won't resolve if the server can't be reached (e.g. offline) or if it returns an internal error.
   *
   * @example
   * ```ts
   * const isAuthenticated  = await nhost.auth.isAuthenticatedAsync();
   *
   * if (isAuthenticated) {
   *   console.log('User is authenticated');
   * }
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/is-authenticated-async
   */
  async isAuthenticatedAsync() {
    return (await this.waitUntilReady()).getSnapshot().matches({ authentication: "signedIn" });
  }
  /**
   * Use `nhost.auth.getAuthenticationStatus` to get the authentication status of the user.
   *
   * If `isLoading` is `true`, the client doesn't know whether the user is authenticated yet or not
   * because some internal authentication network requests have not been resolved yet.
   *
   * The `connectionAttempts` returns the number of times the client has tried to connect to the server with no success (offline, or the server retruned an internal error).
   *
   * @example
   * ```ts
   * const { isAuthenticated, isLoading } = nhost.auth.getAuthenticationStatus();
   *
   * if (isLoading) {
   *   console.log('Loading...')
   * }
   *
   * if (isAuthenticated) {
   *   console.log('User is authenticated');
   * }
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/get-authentication-status
   */
  getAuthenticationStatus() {
    var r;
    const e = ((r = this.client.interpreter) == null ? void 0 : r.getSnapshot().context.importTokenAttempts) || 0;
    return this.isReady() ? { isAuthenticated: this.isAuthenticated(), isLoading: !1, connectionAttempts: e } : {
      isAuthenticated: !1,
      isLoading: !0,
      connectionAttempts: e
    };
  }
  /**
   * Use `nhost.auth.getAccessToken` to get the access token of the user.
   *
   * @example
   * ```ts
   * const accessToken = nhost.auth.getAccessToken();
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/get-access-token
   */
  getAccessToken() {
    var e, r;
    return (r = (e = this._client.interpreter) == null ? void 0 : e.getSnapshot().context.accessToken.value) != null ? r : void 0;
  }
  /**
   * Use `nhost.auth.getDecodedAccessToken` to get the decoded access token of the user.
   *
   * @example
   * ```ts
   * const decodedAccessToken = nhost.auth.getDecodedAccessToken();
   * ```
   *
   * @see {@link https://hasura.io/docs/latest/graphql/core/auth/authentication/jwt/| Hasura documentation}
   * @docs https://docs.nhost.io/reference/javascript/auth/get-decoded-access-token
   */
  getDecodedAccessToken() {
    const e = this.getAccessToken();
    return e ? ce(e) : null;
  }
  /**
   * Use `nhost.auth.getHasuraClaims` to get the Hasura claims of the user.
   *
   * @example
   * ```ts
   * const hasuraClaims = nhost.auth.getHasuraClaims();
   * ```
   *
   * @see {@link https://hasura.io/docs/latest/graphql/core/auth/authentication/jwt/| Hasura documentation}
   * @docs https://docs.nhost.io/reference/javascript/auth/get-hasura-claims
   */
  getHasuraClaims() {
    var e;
    return ((e = this.getDecodedAccessToken()) == null ? void 0 : e["https://hasura.io/jwt/claims"]) || null;
  }
  /**
   * Use `nhost.auth.getHasuraClaim` to get the value of a specific Hasura claim of the user.
   *
   * @example
   * ```ts
   * // if `x-hasura-company-id` exists as a custom claim
   * const companyId = nhost.auth.getHsauraClaim('company-id')
   * ```
   *
   * @param name Name of the variable. You don't have to specify `x-hasura-`.
   *
   * @see {@link https://hasura.io/docs/latest/graphql/core/auth/authentication/jwt/| Hasura documentation}
   * @docs https://docs.nhost.io/reference/javascript/auth/get-hasura-claim
   */
  getHasuraClaim(e) {
    var r;
    return ((r = this.getHasuraClaims()) == null ? void 0 : r[e.startsWith("x-hasura-") ? e : `x-hasura-${e}`]) || null;
  }
  /**
   *
   * Use `nhost.auth.refreshSession` to refresh the session with either the current internal refresh token or an external refresh token.
   *
   * Note: The Nhost client automatically refreshes the session when the user is authenticated but `nhost.auth.refreshSession` can be useful in some special cases.
   *
   * @example
   * ```ts
   * // Refresh the session with the the current internal refresh token.
   * nhost.auth.refreshToken();
   *
   * // Refresh the session with an external refresh token.
   * nhost.auth.refreshToken(refreshToken);
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/refresh-session
   */
  async refreshSession(e) {
    try {
      const r = await this.waitUntilReady();
      return new Promise((t) => {
        const s = e || r.getSnapshot().context.refreshToken.value;
        if (!s)
          return t({ session: null, error: ge });
        const { changed: i } = r.send("TRY_TOKEN", { token: s });
        if (!i)
          return t({ session: null, error: pe });
        r.onTransition((u) => {
          u.matches({ token: { idle: "error" } }) ? t({
            session: null,
            // * TODO get the error from xstate once it is implemented
            error: ye
          }) : u.event.type === "TOKEN_CHANGED" && t({ session: x(u.context), error: null });
        });
      });
    } catch (r) {
      return { session: null, error: r.message };
    }
  }
  /**
   *
   * Use `nhost.auth.getSession()` to get the session of the user.
   *
   * @example
   * ```ts
   * const session = nhost.auth.getSession();
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/get-session
   */
  getSession() {
    var e, r;
    return x((r = (e = this._client.interpreter) == null ? void 0 : e.getSnapshot()) == null ? void 0 : r.context);
  }
  /**
   *
   * Use `nhost.auth.getUser()` to get the signed-in user.
   *
   * @example
   * ```ts
   * const user = nhost.auth.getUser();
   * ```
   *
   * @docs https://docs.nhost.io/reference/javascript/auth/get-user
   */
  getUser() {
    var e, r, t;
    return ((t = (r = (e = this._client.interpreter) == null ? void 0 : e.getSnapshot()) == null ? void 0 : r.context) == null ? void 0 : t.user) || null;
  }
  /**
   * Make sure the state machine is set, and wait for it to be ready
   * @returns
   */
  waitUntilReady() {
    const r = this._client.interpreter;
    if (!r)
      throw Error("Auth interpreter not set");
    return r.getSnapshot().hasTag("loading") ? new Promise((t, s) => {
      let i = setTimeout(
        () => s("The state machine is not yet ready after 15 seconds."),
        15e3
      );
      r.onTransition((u) => {
        if (!u.hasTag("loading"))
          return clearTimeout(i), t(r);
      });
    }) : Promise.resolve(r);
  }
  isReady() {
    var e, r;
    return !((r = (e = this._client.interpreter) == null ? void 0 : e.getSnapshot()) != null && r.hasTag("loading"));
  }
  get client() {
    return this._client;
  }
}
export {
  oe as AuthClient,
  cr as AuthClientSSR,
  $e as AuthCookieClient,
  U as CodifiedError,
  we as EMAIL_NEEDS_VERIFICATION,
  dr as HasuraAuthClient,
  G as INITIAL_MACHINE_CONTEXT,
  I as INVALID_EMAIL_ERROR,
  he as INVALID_MFA_CODE_ERROR,
  me as INVALID_MFA_TICKET_ERROR,
  fe as INVALID_MFA_TYPE_ERROR,
  $ as INVALID_PASSWORD_ERROR,
  Y as INVALID_PHONE_NUMBER_ERROR,
  ye as INVALID_REFRESH_TOKEN,
  Se as INVALID_SIGN_IN_METHOD,
  le as MIN_PASSWORD_LENGTH,
  Z as NETWORK_ERROR_CODE,
  R as NHOST_JWT_EXPIRES_AT_KEY,
  N as NHOST_REFRESH_TOKEN_ID_KEY,
  O as NHOST_REFRESH_TOKEN_KEY,
  Ee as NO_MFA_TICKET_ERROR,
  ge as NO_REFRESH_TOKEN,
  J as OTHER_ERROR_CODE,
  H as REFRESH_TOKEN_MAX_ATTEMPTS,
  C as STATE_ERROR_CODE,
  pe as TOKEN_REFRESHER_RUNNING_ERROR,
  de as TOKEN_REFRESH_MARGIN_SECONDS,
  S as USER_ALREADY_SIGNED_IN,
  or as USER_NOT_ANONYMOUS,
  Te as USER_UNAUTHENTICATED,
  y as VALIDATION_ERROR_CODE,
  lr as activateMfaPromise,
  qe as addSecurityKeyPromise,
  We as changeEmailPromise,
  je as changePasswordPromise,
  Me as createAuthMachine,
  Ke as createChangeEmailMachine,
  Ve as createChangePasswordMachine,
  ar as createEnableMfaMachine,
  Ge as createResetPasswordMachine,
  Le as createSendVerificationEmailMachine,
  ie as encodeQueryParameters,
  ur as generateQrCodePromise,
  A as getAuthenticationResult,
  De as getFetch,
  D as getParameterByName,
  x as getSession,
  q as isBrowser,
  v as isValidEmail,
  W as isValidPassword,
  z as isValidPhoneNumber,
  Ue as isValidTicket,
  Ce as localStorageGetter,
  Ne as localStorageSetter,
  k as postFetch,
  F as removeParameterFromWindow,
  He as resetPasswordPromise,
  w as rewriteRedirectTo,
  Ye as sendVerificationEmailPromise,
  Fe as signInAnonymousPromise,
  ze as signInEmailPasswordPromise,
  Q as signInEmailPasswordlessPromise,
  Qe as signInEmailSecurityKeyPromise,
  Be as signInMfaTotpPromise,
  Xe as signInPATPromise,
  Ze as signInSmsPasswordlessOtpPromise,
  B as signInSmsPasswordlessPromise,
  Je as signOutPromise,
  X as signUpEmailPasswordPromise,
  er as signUpEmailSecurityKeyPromise
};
